### PR TITLE
[reddit] Support gallery playlist download

### PR DESCRIPTION
**Pull request by sethtroisi**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] [Searched](https://github.com/ytdl-org/youtube-dl/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Read [adding new extractor tutorial](https://github.com/ytdl-org/youtube-dl#adding-support-for-a-new-site)
- [x] Read [youtube-dl coding conventions](https://github.com/ytdl-org/youtube-dl#youtube-dl-coding-conventions) and adjusted the code to meet them
- [ ] Covered the code with tests (note that PRs without tests will be REJECTED)
  - I'm getting errors here, see comments in code
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [x] Improvement
- [ ] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

Support extraction of Reddit gallery "playlists".

This works both for raw gallery urls[1] and posts which get resolved to galleries [2]

[1] https://www.reddit.com/gallery/j8foda/
[2] https://www.reddit.com/r/BostonTerrier/comments/j8foda/my_mookie/

### CODE REVIEW

File: youtube_dl/extractor/reddit.py [hunk 1]
```diff
@@ -49,6 +51,58 @@ def _real_extract(self, url):
...
+            'title': 'My Mookie',
+            'entries': [
+                {'url': 'https://i.reddit.com/flmjm893p7s51.jpg'},
+                {'url': 'https://i.reddit.com/fhcwp0793p7s5.jpg'},
+            ]
+        },
+        'params': {
+            'format': 'bestvideo',
+        },
+    }
```
Comment (sethtroisi) [inline youtube_dl/extractor/reddit.py:71]: This test is breaking and searching other extractors I wasn't able to figure out what I need to do.
I'd appreciate pointers to documentation/other extractors.

The error from `python test/test_download.py -v TestDownload.test_RedditG` is

```
test_RedditG (__main__.TestDownload): ... [RedditG] j8foda: Downloading JSON metadata
[download] Downloading playlist: j8foda
[RedditG] playlist j8foda: Collected 2 video ids (downloading 2 of them)
[download] Downloading video 1 of 2
[generic] flmjm893p7s51: Requesting header
[generic] flmjm893p7s51: Downloading webpage
ERROR: URL could be a direct video link, returning it as such.; please report this issue on https://yt-dl.org/bug . Make sure you are using the latest version; see  https://yt-dl.org/update  on how to update. Be sure to call youtube-dl with the --verbose flag and include its complete output.
Traceback (most recent call last):
  File "/home/five/Projects/youtube-dl/youtube_dl/YoutubeDL.py", line 806, in wrapper
    return func(self, *args, **kwargs)
  File "/home/five/Projects/youtube-dl/youtube_dl/YoutubeDL.py", line 827, in __extract_info
    ie_result = ie.extract(url)
  File "/home/five/Projects/youtube-dl/youtube_dl/extractor/common.py", line 532, in extract
    ie_result = self._real_extract(url)
  File "/home/five/Projects/youtube-dl/youtube_dl/extractor/generic.py", line 2479, in _real_extract
    self._downloader.report_warning(
  File "/home/five/Projects/youtube-dl/test/helper.py", line 271, in _report_warning
    real_warning(w)
  File "test/test_download.py", line 52, in report_warning
    raise ExtractorError(message)
youtube_dl.utils.ExtractorError: URL could be a direct video link, returning it as such.; please report this issue on https://yt-dl.org/bug . Make sure you are using the latest version; see  https://yt-dl.org/update  on how to update. Be sure to call youtube-dl with the --verbose flag and include its complete output.

ERROR
```

Which suggests I need to mark my entries as direct download links?

File: youtube_dl/extractor/reddit.py [hunk 2]
```diff
@@ -49,6 +51,58 @@ def _real_extract(self, url):
...
+
+                    # Need to extract fileextension from url
+                    filename = url_basename(data['s']['u'])
+                    assert filename.startswith(key), (key, filename)
+
+                    entry = self.url_result(
+                        urljoin('https://i.redd.it', filename),
+                        ie=None,
+                        video_id=key,
+                        video_title=filename)
```
Comment (sethtroisi) [inline youtube_dl/extractor/reddit.py:96]: These parameters don't affect the downloaded filename (probably because of the direct download problem.

**Comment by remitamine** [comment]
youtube-dl only supports downloading video/audio files.

**Comment by sethtroisi** [comment]
I'll re-open when Reddit galleries support videos.

**Pull request closed without merging**